tempesta_fw/http.c

```diff
@@ -615,35 +634,212 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
 	return tfw_http_set_hdr_connection(m, conn_flg);
 }
 
+static inline bool
+tfw_http_req_is_idempotent(TfwHttpReq *req)
+{
+	return (req->method != TFW_HTTP_METH_POST);
+}
+
 /*
- * Depending on results of processing of a request, either send the request
- * to an appropriate server, or return the cached response. If none of that
- * can be done for any reason, return HTTP 404 or 500 error to the client.
+ * Forward request @req to server connection @srv_conn.
+ * Can be called in both SoftIRQ and kernel thread contexts.
+ */
+static void
+tfw_http_conn_req_forward(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	TFW_DBG3("%s: srv_conn = %p, req = %p\n", __func__, srv_conn, req);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	spin_lock_bh(&srv_conn->msg_qlock);
+	if (!tfw_http_req_is_idempotent(req))
+		req->flags |= TFW_HTTP_SYNC_REQ;
+	list_add_tail(&req->msg.fwd_list, &srv_conn->msg_queue);
+	if (srv_conn->flags & TFW_CONN_FWD_HOLD) {
+		spin_unlock_bh(&srv_conn->msg_qlock);
+		return;
+	}
+	if (tfw_connection_send(srv_conn, (TfwMsg *)req, false)) {
+		list_del_init(&req->msg.fwd_list);
+		spin_unlock_bh(&srv_conn->msg_qlock);
+		tfw_http_send_500((TfwHttpMsg *)req);
+		TFW_INC_STAT_BH(clnt.msgs_otherr);
+		return;
+	}
+	if (req->flags & TFW_HTTP_SYNC_REQ)
+		srv_conn->flags |= TFW_CONN_FWD_HOLD;
+	spin_unlock_bh(&srv_conn->msg_qlock);
+	TFW_INC_STAT_BH(clnt.msgs_forwarded);
+}
+
+/*
+ * Forward stalled requests in server connection @srv_conn.
+ * Can be called in both SoftIRQ and kernel thread contexts.
+ */
+static void
+tfw_http_conn_req_forward_stalled(TfwConnection *srv_conn)
+{
+	TfwHttpReq *req, *tmp, *end;
+	struct list_head zap_queue, err_queue;
+	struct list_head *fwd_queue = &srv_conn->msg_queue;
+
+	TFW_DBG3("%s: conn = %p\n", __func__, srv_conn);
+	BUG_ON(!(srv_conn->flags & TFW_CONN_FWD_HOLD));
+
+	INIT_LIST_HEAD(&zap_queue);
+	INIT_LIST_HEAD(&err_queue);
+
+	spin_lock_bh(&srv_conn->msg_qlock);
+	end = container_of(fwd_queue, TfwHttpReq, msg.fwd_list);
+	list_for_each_entry_safe(req, tmp, fwd_queue, msg.fwd_list) {
```

Comment:
`srv_conn->msg_queue` contains messages sent to the server and waiting for response. It seems we should process new `backlog` server queue, which contains backlogged messages while there is nonidempotent message is in `msg_queue`, here.